File: frontend/src/types/queryBuilder.ts
export interface FilterCondition {
  dimension_id: string;
  operator: '=' | '!=' | '>' | '<' | '>=' | '<=' | 'IN' | 'NOT IN' | 'LIKE' | 'IS NULL' | 'IS NOT NULL';
  value?: any;
}

export interface QueryRequest {
  entity_id: string;
  dimension_ids: string[];
  measure_ids: string[];
  filters?: FilterCondition[];
  limit?: number;
}

export interface QueryResponse {
  columns: string[];
  data: Record<string, any>[];
  row_count: number;
  generated_sql: string;
  entity_name: string;
}

// Saved Query Types
export interface SavedQuery {
  id: string;
  owner_id: string;
  name: string;
  description?: string;
  entity_id: string;
  query_config: {
    dimension_ids: string[];
    measure_ids: string[];
    filters: FilterCondition[];
    limit: number;
  };
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface SavedQueryCreate {
  name: string;
  description?: string;
  entity_id: string;
  query_config: {
    dimension_ids: string[];
    measure_ids: string[];
    filters: FilterCondition[];
    limit: number;
  };
  is_favorite?: boolean;
}

export interface SavedQueryUpdate {
  name?: string;
  description?: string;
  query_config?: {
    dimension_ids: string[];
    measure_ids: string[];
    filters: FilterCondition[];
    limit: number;
  };
  is_favorite?: boolean;
}

export interface SavedQueryListResponse {
  queries: SavedQuery[];
  total: number;
}

// Query History Types
export interface QueryHistory {
  id: string;
  owner_id: string;
  entity_id: string;
  query_config: {
    dimension_ids: string[];
    measure_ids: string[];
    filters: FilterCondition[];
    limit: number;
  };
  generated_sql: string;
  row_count: number;
  execution_time_ms: number;
  executed_at: string;
}

export interface QueryHistoryListResponse {
  history: QueryHistory[];
  total: number;
}

File: backend/app/schemas/query_builder.py
"""Query Builder schemas for request and response validation."""
from typing import List, Optional, Any
from uuid import UUID
from pydantic import BaseModel, Field


class FilterCondition(BaseModel):
    """Filter condition for query."""
    dimension_id: UUID
    operator: str = Field(
        ...,
        pattern="^(=|!=|>|<|>=|<=|IN|NOT IN|LIKE|IS NULL|IS NOT NULL)$"
    )
    value: Optional[Any] = None  # Can be single value or list for IN operator


class QueryRequest(BaseModel):
    """Request schema for query execution."""
    entity_id: UUID
    dimension_ids: List[UUID] = Field(
        default_factory=list,
        description="Dimensions for GROUP BY"
    )
    measure_ids: List[UUID] = Field(
        ...,
        min_length=1,
        description="Measures to calculate"
    )
    filters: List[FilterCondition] = Field(default_factory=list)
    limit: Optional[int] = Field(
        default=1000,
        le=10000,
        description="Row limit"
    )


class QueryResponse(BaseModel):
    """Response schema for query results."""
    columns: List[str]
    data: List[dict]  # Each row as a dictionary
    row_count: int
    generated_sql: str
    entity_name: str
